addon/-private/validated-changeset.ts

```diff
@@ -0,0 +1,933 @@
+   * errors.
+   *
+   * @method rollback
+   */
+  rollback(): IChangeset {
+    // Get keys before reset.
+    let c: BufferedChangeset = this;
+    let keys = c._rollbackKeys();
+
+    // Reset.
+    this[CHANGES] = {};
+    this[ERRORS] = {};
+    c._notifyVirtualProperties(keys)
+
+    c.trigger(AFTER_ROLLBACK_EVENT);
+    return this;
+  }
+
+  /**
+   * Discards any errors, keeping only valid changes.
+   *
+   * @public
+   * @chainable
+   * @method rollbackInvalid
+   * @param {String} key optional key to rollback invalid values
+   * @return {Changeset}
+   */
+  rollbackInvalid(key: string | void): IChangeset {
+    let errorKeys = keys(this[ERRORS]);
+
+    if (key) {
+      this._notifyVirtualProperties([key]);
+      // @tracked
+      this[ERRORS] = this._deleteKey(ERRORS, key) as Errors<any>;
+      if (errorKeys.indexOf(key) > -1) {
+        this[CHANGES] = this._deleteKey(CHANGES, key) as Changes;
+      }
+    } else {
+      this._notifyVirtualProperties();
+      this[ERRORS] = {};
+
+      // if on CHANGES hash, rollback those as well
+      errorKeys.forEach((errKey) => {
+        this[CHANGES] = this._deleteKey(CHANGES, errKey) as Changes;
+      })
+    }
+
+    return this;
+  }
+
+  /**
+   * Discards changes/errors for the specified properly only.
+   *
+   * @public
+   * @chainable
+   * @method rollbackProperty
+   * @param {String} key key to delete off of changes and errors
+   * @return {Changeset}
+   */
+  rollbackProperty(key: string): IChangeset {
+    // @tracked
+    this[CHANGES] = this._deleteKey(CHANGES, key) as Changes;
+    // @tracked
+    this[ERRORS] = this._deleteKey(ERRORS, key) as Errors<any>;
+
+    return this;
+  }
+
+  /**
+   * Validates the changeset immediately against the validationMap passed in.
+   * If no key is passed into this method, it will validate all fields on the
+   * validationMap and set errors accordingly. Will throw an error if no
+   * validationMap is present.
+   *
+   * @method validate
+   */
+  validate(
+    key?: string | undefined
+  ): Promise<null> | Promise<any | IErr<any>> | Promise<Array<any | IErr<any>>> {
+    if (keys(this.validationMap as object).length === 0) {
+      return Promise.resolve(null);
+    }
+
+    if (!Boolean(key)) {
+      let maybePromise = keys(this.validationMap as object).map(validationKey => {
+        return this._validateKey(validationKey, this._valueFor(validationKey));
+      });
+
+      return Promise.all(maybePromise);
+    }
+
+    return Promise.resolve(this._validateKey(<string>key, this._valueFor(<string>key)));
+  }
+
+  /**
+   * Manually add an error to the changeset. If there is an existing
+   * error or change for `key`, it will be overwritten.
+   *
+   * @method addError
+   */
+  addError<T> (
+    key: string,
+    error: IErr<T> | ValidationErr
+  ): IErr<T> | ValidationErr {
+    // Construct new `Err` instance.
+    let newError;
+    if (isObject(error) && !Array.isArray(error)) {
```

Comment:
This looks like a type guard. You can do:

```ts
const isIErr = <T>(error: unknown): error is IErr<T> => isObject(error) && !Array.isArray(error);
```

And you won't need to do a cast later in the branch.